README.md
Limit Login Attempts (module for Omeka S)
=========================================

[Limit Login Attempts] is a module for [Omeka S] that limits the rate of login
attempts for each IP (not by auth cookie) in order to avoid brute-force attacks.

This module is a full rewrite of a [plugin for WordPress], create by Johan Eenfeldt (johanee).


Description
-----------

Limit the number of login attempts possible both through normal login.

By default Omeka S allows unlimited login attempts either through the login
page. This allows passwords (or hashes) to be brute-force cracked with relative
ease.

Limit Login Attempts blocks an Internet address from making further attempts
after a specified limit on retries is reached, making a brute-force attack
difficult or impossible.

## Features

* Limit the number of retry attempts when logging in (for each IP)
* Fully customizable
* Informs user about remaining retries or lockout time on login page
* Optional logging and optional email notification
* Handles server behind reverse proxy
* It is possible to whitelist IPs, but you probably shouldn't. :-)

Translations: Bulgarian, Brazilian Portuguese, Catalan, Chinese (Traditional),
Czech, Dutch, Finnish, French, German, Hungarian, Norwegian, Persian, Romanian,
Russian, Spanish, Swedish, Turkish.

The module uses standard actions and filters only.

## Screenshots

1. Login screen after failed login with retries remaining.

  ![Login screen after failed login](https://github.com/Daniel-KM/Omeka-S-module-LimitLoginAttempts/blob/master/data/readme/lockout_attempt.png)

2. Login screen during lockout.

  ![Login screen during lockout](https://github.com/Daniel-KM/Omeka-S-module-LimitLoginAttempts/blob/master/data/readme/lockout_blocked.png)

3. Administration interface in Omeka S

  ![administration interface](https://github.com/Daniel-KM/Omeka-S-module-LimitLoginAttempts/blob/master/data/readme/lockout_config.png)


Installation
------------

Uncompress files in the module directory and rename module folder `LimitLoginAttempts`.

Then install it like any other Omeka module and follow the config instructions.

If your server is located behind a reverse proxy, make sure to set the option.


Frequently Asked Questions
--------------------------

## Why not reset failed attempts on a successful login?

This is very much by design. Otherwise you could brute force the "admin"
password by logging in as your own user every 4th attempt.

## What is this option about site connection and reverse proxy?

A reverse proxy is a server in between the site and the Internet (perhaps
handling caching or load-balancing). This makes getting the correct client IP to
block slightly more complicated.

The option default to NOT being behind a proxy -- which should be by far the
common case.

## How do I know if my site is behind a reverse proxy?

You probably are not or you would know. We show a pretty good guess on the
option page. Set the option using this unless you are sure you know better.

## Can I whitelist my IP so I don't get locked out?

First please consider if you really need this. Generally speaking it is not a
good idea to have exceptions to your security policies.

Note that we still do notification and logging as usual. This is meant to allow
you to be aware of any suspicious activity from whitelisted IPs.

## I locked myself out testing this thing, what do I do?

Either wait, or:

If you know how to edit / add to PHP files you can use the IP whitelist
functionality described above. You should then use the "Restore Lockouts" button
on the module settings page and remove the whitelist function again.

If you have ftp / ssh access to the site, remove the folder of the module or
increase the version number in the `config/module.ini`, so it will deactivate it.

If you have access to the database (for example through phpMyAdmin) you can clear
the limit_login_lockouts option in the Omeka S `setting` table. The sql for a
standard install is: `UPDATE setting SET value = '' WHERE id = 'limit_login_lockouts';`
You can disable the module too: `UPDATE module SET is_active = 0 WHERE id = 'LimitLoginAttempts';`.


Warning
-------

Use it at your own risk.

It?s always recommended to backup your files and your databases and to check
your archives regularly so you can roll back if needed.


Troubleshooting
---------------

See online issues on the [module issues] page on GitHub.


License
-------

This module is published under the [GNU/GPL] license.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


Contacts
--------

* Daniel Berthereau (see [Daniel-KM] on GitHub)


Copyright
---------

* Copright Johan Eenfeldt, 2008-2012
* Copright Daniel Berthereau, 2017
* Translations: see the [WordPress page]

Thanks to Michael Skerwiderski for reverse proxy handling suggestions (WordPress).


[Limit Login Attempts]: https://github.com/Daniel-KM/Omeka-S-module-LimitLoginAttempts
[Omeka S]: https://omeka.org/s
[plugin for WordPress]: https://wordpress.org/plugins/limit-login-attempts
[module issues]: https://github.com/Daniel-KM/Omeka-S-module-LimitLoginAttempts/issues
[GNU/GPL]: https://www.gnu.org/licenses/gpl-3.0.html
[WordPress page]: https://translate.wordpress.org/projects/wp-plugins/limit-login-attempts/contributors
[Daniel-KM]: https://github.com/Daniel-KM "Daniel Berthereau"


Changelog.md
Changelog
=========

= 3.2 =
Ported to Omeka S and renamed "Lockout". See commits for next logs.

= 1.7.1 =
This version fixes a security bug in version 1.6.2 and 1.7.0. Please upgrade immediately.

"Auth cookies" are special cookies set at login that authenticating you to the system. It is how WordPress "remembers" that you are logged in between page loads.

During lockout these are supposed to be cleared, but a change in 1.6.2 broke this. It allowed an attacker to keep trying to break these cookies during a lockout.

Lockout of normal password login attempts still worked as it should, and it appears that all "auth cookie" attempts would keep getting logged.

In theory the "auth cookie" is quite resistant to brute force attack. It contains a cryptographic hash of the user password, and the difficulty to break it is not based on the password strength but instead on the cryptographic operations used and the length of the hash value. In theory it should take many many years to break this hash. As theory and practice does not always agree it is still a good idea to have working lockouts of any such attempts.

= 1.7.0 =
* Added filter that allows whitelisting IP. Please use with care!!
* Update to Spanish translation, thanks to Marcelo Pedra
* Updated Swedish translation
* Tested against WordPress 3.3.2

= 1.6.2 =
* Fix bug where log would not get updated after it had been cleared
* Do plugin setup in 'init' action
* Small update to Spanish translation file, thanks to Marcelo Pedra
* Tested against WordPress 3.2.1

= 1.6.1 =
* (WordPress 3.0+) An invalid cookie can sometimes get sent multiple times before it gets cleared, resulting in multiple failed attempts or even a lockout from a single invalid cookie. Store the latest failed cookie to make sure we only count it as one failed attempt
* Define "Text Domain" correctly
* Include correct Dutch tranlation file. Thanks to Martin1 for noticing. Thanks again to Bjorn Wijers for the translation
* Updated POT file for this version
* Tested against WordPress 3.1-RC4

= 1.6.0 =
* Happy New Year
* Tested against WordPress 3.1-RC1
* Plugin now requires WordPress version 2.8+. Of course you should never ever use anything but the latest version
* Fixed deprecation warnings that had been piling up with the old version requirement. Thanks to Johannes Ruthenberg for the report that prompted this
* Removed auth cookie admin check for version 2.7.
* Make sure relevant values in $_COOKIE get cleared right away on auth cookie validation failure. There are still some problems with cookie auth handling. The lockout can trigger prematurely in rare cases, but fixing it is plugin version 2 stuff unfortunately.
* Changed default time for retries to reset from 24 hours to 12 hours. The security impact is very minor and it means the warning will disappear "overnight"
* Added question to FAQ ("Why not reset failed attempts on a successful login?")
* Updated screenshots

= 1.5.2 =
* Reverted minor cookie-handling cleanup which might somehow be responsible for recently reported cookie related lockouts
* Added version 1.x Brazilian Portuguese translation, thanks to Luciano Passuello
* Added Finnish translation, thanks to Ari Kontiainen

= 1.5.1 =
* Further multisite & WPMU support (again thanks to <erik@erikshosting.com>)
* Better error handling if option variables are damaged
* Added Traditional Chinese translation, thanks to Denny Huang <bigexplorations@bigexplorations.com.tw>

= 1.5 =
* Tested against WordPress 3.0
* Handle 3.0 login page failure "shake"
* Basic multisite support (parts thanks to <erik@erikshosting.com>)
* Added Dutch translation, thanks to Bjorn Wijers <burobjorn@burobjorn.nl>
* Added Hungarian translation, thanks to Bálint Vereskuti <balint@vereskuti.info>
* Added French translation, thanks to oVa <ova13lastar@gmail.com>

= 1.4.1 =
* Added Turkish translation, thanks to Yazan Canarkadas

= 1.4 =
* Protect admin page update using wp_nonce
* Added Czech translation, thanks to Jakub Jedelsky

= 1.3.2 =
* Added Bulgarian translation, thanks to Hristo Chakarov
* Added Norwegian translation, thanks to Rune Gulbrandsøy
* Added Spanish translation, thanks to Marcelo Pedra
* Added Persian translation, thanks to Mostafa Soufi
* Added Russian translation, thanks to Jack Leonid (http://studio-xl.com)

= 1.3.1 =
* Added Catalan translation, thanks to Robert Buj
* Added Romanian translation, thanks to Robert Tudor

= 1.3 =
* Support for getting the correct IP for clients while server is behind reverse proxy, thanks to Michael Skerwiderski
* Added German translation, thanks to Michael Skerwiderski

= 1.2 =
* No longer replaces pluggable function when cookie handling active. Re-implemented using available actions and filters
* Filter error messages during login to avoid information leak regarding available usernames
* Do not show retries or lockout messages except for login (registration, lost password pages). No change in actual enforcement
* Slightly more aggressive in trimming old retries data

= 1.1 =
* Added translation support
* Added Swedish translation
* During lockout, filter out all other login errors
* Minor cleanups

= 1.0 =
* Initial version

== Upgrade Notice ==

= 1.7.1 =
Users of version 1.6.2 and 1.7.0 should upgrade immediately. There was a problem with "auth cookie" lockout enforcement. Lockout of normal password login attempts still worked as it should. Please see plugin Changelog for more information.


src/Form/Config.php
<?php
namespace LimitLoginAttempts\Form;

use Zend\Form\Form;

class Config extends Form
{
    public function init()
    {
        // Resets.

        $this->add([
            'name' => 'limit_login_clear_total_lockouts',
            'type' => 'Checkbox',
            'options' => [
                'label' => 'Reset total lockouts', // @translate
            ],
            'attributes' => [
                'value' => false,
            ],
        ]);

        $this->add([
            'name' => 'limit_login_clear_current_lockouts',
            'type' => 'Checkbox',
            'options' => [
                'label' => 'Clear current lockouts', // @translate
            ],
            'attributes' => [
                'value' => false,
            ],
        ]);

        $this->add([
            'name' => 'limit_login_clear_log',
            'type' => 'Checkbox',
            'options' => [
                'label' => 'Clear IP log', // @translate
            ],
            'attributes' => [
                'value' => false,
            ],
        ]);

        // Lockout.

        $this->add([
            'name' => 'limit_login_allowed_retries',
            'type' => 'Text',
            'options' => [
                'label' => 'Allowed retries', // @translate
            ],
        ]);

        $this->add([
            'name' => 'limit_login_lockout_duration',
            'type' => 'Text',
            'options' => [
                'label' => 'Lockout duration (seconds)', // @translate
            ],
        ]);

        $this->add([
            'name' => 'limit_login_allowed_lockouts',
            'type' => 'Text',
            'options' => [
                'label' => 'Allowed retries', // @translate
            ],
        ]);

        $this->add([
            'name' => 'limit_login_long_duration',
            'type' => 'Text',
            'options' => [
                'label' => 'Long lock out (seconds)', // @translate
            ],
        ]);

        $this->add([
            'name' => 'limit_login_valid_duration',
            'type' => 'Text',
            'options' => [
                'label' => 'Valid duration (seconds)', // @translate
            ],
        ]);

        // Site connection.

        $this->add([
            'type' => 'Radio',
            'name' => 'limit_login_client_type',
            'options' => [
                'label' => 'Client type', // @translate
                'value_options' => [
                    \LimitLoginAttempts\Module::DIRECT_ADDR => 'Direct connection', // @translate
                    \LimitLoginAttempts\Module::PROXY_ADDR => 'From behind a reversy proxy', // @translate
                ],
            ],
        ]);

        $this->add([
            'name' => 'limit_login_cookies',
            'type' => 'Checkbox',
            'options' => [
                'label' => 'Handle cookie login', // @translate
            ],
        ]);

        // Notification.

        $this->add([
            'name' => 'limit_login_lockout_notify',
            'type' => 'MultiCheckbox',
            'options' => [
                'label' => 'Notify on lockout', // @translate
                'value_options' => [
                    'log' => 'Log IP', // @translate
                    'email' => 'Email', // @translate
                ],
            ],
        ]);

        $this->add([
            'name' => 'limit_login_notify_email_after',
            'type' => 'Text',
            'options' => [
                'label' => 'Email to admin after attempts', // @translate
            ],
        ]);
    }
}


config/module.config.php
<?php
namespace LimitLoginAttempts;

return [
    'view_manager' => [
        'template_path_stack' => [
            OMEKA_PATH . '/modules/LimitLoginAttempts/view',
        ],
    ],
    'controllers' => [
        'factories' => [
            // Override the standard Omeka login controller.
            'Omeka\Controller\Login' => Service\Controller\LoginControllerFactory::class,
        ],
    ],
    'form_elements' => [
        'invokables' => [
            'LimitLoginAttempts\Form\Config' => Form\Config::class,
        ],
    ],
    'translator' => [
        'translation_file_patterns' => [
            [
                'type' => 'gettext',
                'base_dir' => __DIR__ . '/../language',
                'pattern' => '%s.mo',
                'text_domain' => null,
            ],
        ],
    ],
];


Module.php
<?php
namespace LimitLoginAttempts;

/*
 * Copyright Johan Eenfeldt, 2008-2012
 * Copyright Daniel Berthereau, 2017
 *
 * Licenced under the GNU GPL:
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

use LimitLoginAttempts\Form\Config as ConfigForm;
use Omeka\Module\AbstractModule;
use Zend\Mvc\Controller\AbstractController;
use Zend\ServiceManager\ServiceLocatorInterface;
use Zend\View\Renderer\PhpRenderer;

/**
 * Limit Login Attempts
 *
 * Limit rate of login attempts, including by way of cookies, for each IP.
 *
 * @copyright Johan Eenfeldt, 2008-2012
 * @copyright Daniel Berthereau, 2017
 * @license Gnu/Gpl v3
 */
class Module extends AbstractModule
{
    const DIRECT_ADDR = 'REMOTE_ADDR';
    const PROXY_ADDR = 'HTTP_X_FORWARDED_FOR';

    /**
     * Settings and their default values.
     *
     * @var array
     */
    protected $settings = [
        // Are we behind a proxy?
        'limit_login_client_type' => self::DIRECT_ADDR,

        // Lock out after this many tries.
        'limit_login_allowed_retries' => 4,
        // Lock out for this many seconds (default is 20 minutes).
        'limit_login_lockout_duration' => 1200,
        // Long lock out after this many lockouts.
        'limit_login_allowed_lockouts' => 4,
        // Long lock out for this many seconds (default is 24 hours).
        'limit_login_long_duration' => 86400,
        // Reset failed attempts after this many seconds (defaul is 12 hours).
        'limit_login_valid_duration' => 43200,

        // Also limit malformed/forged cookies?
        'limit_login_cookies' => true,
        // Notify on lockout. Values: '', 'log' and/or 'email'.
        'limit_login_lockout_notify' => ['log'],
        // If notify by email, do so after this number of lockouts.
        'limit_login_notify_email_after' => 4,

        // Total lockouts.
        'limit_login_lockouts_total' => 0,
        // Current lockouts.
        'limit_login_lockouts' => [],
        // Logs.
        'limit_login_logged' => [],
    ];

    public function getConfig()
    {
        return include __DIR__ . '/config/module.config.php';
    }

    public function install(ServiceLocatorInterface $serviceLocator)
    {
        $settings = $serviceLocator->get('Omeka\Settings');
        foreach ($this->settings as $name => $value) {
            $settings->set($name, $value);
        }
    }

    public function uninstall(ServiceLocatorInterface $serviceLocator)
    {
        $settings = $serviceLocator->get('Omeka\Settings');
        foreach ($this->settings as $name => $value) {
            $settings->delete($name);
        }
    }

    public function getConfigForm(PhpRenderer $renderer)
    {
        $services = $this->getServiceLocator();
        $settings = $services->get('Omeka\Settings');
        $formElementManager = $services->get('FormElementManager');

        // TODO $this->limit_login_cleanup();

        $formData = [];
        foreach ($this->settings as $name => $value) {
            $formData[$name] = $settings->get($name);
        }

        $form = $formElementManager->get(ConfigForm::class);
        $form->init();
        $form->setData($formData);

        $clientTypeGuess = $this->guessProxy();
        if ($clientTypeGuess== self::DIRECT_ADDR) {
            $clientTypeMessage = sprintf('It appears the site is reached directly (from your IP: %s)', // @translate
                '<strong>' . $this->getAddress(self::DIRECT_ADDR) . '</strong>');
        } else {
            $clientTypeMessage = sprintf('It appears the site is reached through a proxy server (proxy IP: %s, your IP: %s)', // @translate
                '<strong>' . $this->getAddress(self::PROXY_ADDR) . '</strong>',
                '<strong>' . $this->getAddress(self::DIRECT_ADDR) . '</strong>');
        }

        // Allow to display fieldsets in config form.
        $vars = [];
        $vars['form'] = $form;

        $vars['lockout_total'] = $settings->get('limit_login_lockout_total', 0);
        $vars['lockouts'] = $settings->get('llimit_login_lockouts', []);
        $vars['client_type_message'] = $clientTypeMessage;
        $vars['client_type_warning'] = $clientTypeGuess != $settings->get('limit_login_client_type', $this->settings['limit_login_client_type']);
        $vars['log'] = $settings->get('limit_login_logged', []);

        return $renderer->render('limit-login-attempts/module/config.phtml', $vars);
    }

    public function handleConfigForm(AbstractController $controller)
    {
        $services = $this->getServiceLocator();
        $settings = $services->get('Omeka\Settings');

        $params = $controller->getRequest()->getPost();

        // $form = new ConfigForm;
        // $form->init();
        // $form->setData($params);
        // if (!$form->isValid()) {
        //     $controller->messenger()->addErrors($form->getMessages());
        //     return false;
        // }

        if (!empty($params['limit_login_clear_total_lockouts'])) {
            $params['limit_login_lockouts_total'] = 0;
            $controller->messenger()->addSuccess('Reset lockout count'); // @translate
        }

        if (!empty($params['limit_login_clear_current_lockouts'])) {
            $params['limit_login_lockouts_total'] = [];
            $controller->messenger()->addSuccess('Cleared current lockouts'); // @translate
        }

        if (!empty($params['limit_login_clear_log'])) {
            $params['limit_login_logged'] = [];
            $controller->messenger()->addSuccess('Cleared IP log'); // @translate
        }

        // Clean params.
        $params['limit_login_allowed_retries'] = (int) $params['limit_login_allowed_retries'];
        $params['limit_login_lockout_duration'] = (int) $params['limit_login_lockout_duration'];
        $params['limit_login_valid_duration'] = (int) $params['limit_login_valid_duration'];
        $params['limit_login_allowed_lockouts'] = (int) $params['limit_login_allowed_lockouts'];
        $params['limit_login_long_duration'] = (int) $params['limit_login_long_duration'];
        $params['limit_login_cookies'] = (bool) $params['limit_login_cookies'];
        $params['limit_login_notify_email_after'] = (int) $params['limit_login_notify_email_after'];
        $params['limit_login_notify_email_after'] = (int) $params['limit_login_notify_email_after'];
        $params['limit_login_lockout_notify'] = array_intersect($params['limit_login_lockout_notify'], ['log', 'email']);
        if (!in_array($params['limit_login_client_type'], [self::DIRECT_ADDR, self::PROXY_ADDR])) {
            $params['limit_login_client_type'] = self::DIRECT_ADDR;
        }

        foreach ($params as $name => $value) {
            if (isset($this->settings[$name])) {
                $settings->set($name, $value);
            }
        }
    }

    /**
     * Get correct remote address.
     */
    function getAddress($typeName = '')
    {
        $type = $typeName;
        if (empty($type)) {
            $type = self::DIRECT_ADDR;
        }

        if (isset($_SERVER[$type])) {
            return $_SERVER[$type];
        }

        // Not found. Did we get proxy type from option?
        // If so, try to fall back to direct address.
        if (empty($type_name) && $type == self::PROXY_ADDR && isset($_SERVER[self::DIRECT_ADDR])) {

            // NOTE: Even though we fall back to direct address -- meaning you
            // can get a mostly working plugin when set to PROXY mode while in
            // fact directly connected to Internet it is not safe!
            //
            // Client can itself send HTTP_X_FORWARDED_FOR header fooling us
            // regarding which IP should be banned.
            return $_SERVER[self::DIRECT_ADDR];
        }

        return '';
    }

    /**
     * Make a guess if we are behind a proxy or not.
     */
    function guessProxy()
    {
        return isset($_SERVER[self::PROXY_ADDR])
            ? self::PROXY_ADDR
            : self::DIRECT_ADDR;
    }
}
